README
NAME
    Type::Tiny::Manual - an overview of Type::Tiny

SYNOPSIS
    Type::Tiny is a small class for writing type constraints, inspired by
    Moose's type constraint API. It has only one non-core dependency (and even
    that is simply a module that was previously distributed as part of
    Type::Tiny but has since been spun off), and can be used with Moose, Mouse
    and Moo (or none of the above).

    Type::Tiny is bundled with Type::Library a framework for organizing type
    constraints into collections.

    Also bundled is Types::Standard, a Moose-inspired library of useful type
    constraints.

    Type::Params is also provided, to allow very fast checking and coercion of
    function and method parameters.

SEE ALSO
    *   Libraries - how to build a type library with Type::Tiny, Type::Library
        and Type::Utils

    *   Coercions - adding coercions to type constraints

    *   Using with Moose - how to use Type::Tiny and Type::Library with Moose

    *   Using with Mouse - how to use Type::Tiny and Type::Library with Mouse

    *   Using with Moo - how to use Type::Tiny and Type::Library with Moo

    *   Using with Other OO Frameworks - how to use Type::Tiny and
        Type::Library with other OO frameworks

    *   Type::Tiny and friends don't need to be used within an OO framework.
        See FreeMind::Node for an example that does not.

    *   Processing arguments to subs - coerce and validate arguments to
        functions and methods.

    *   Other modules using Type::Tiny in interesting ways: Type::Tie,
        Test::Mocha, Scalar::Does, Set::Equivalence...

    *   Optimization - squeeze the most out of your CPU.

    *   Type::Tiny maintenance policies - the stability policy.

DEPENDENCIES
    Type::Tiny requires at least Perl 5.6.1, though certain Unicode-related
    features (e.g. non-ASCII type constraint names) may work better in newer
    versions of Perl.

    Type::Tiny requires Exporter::Tiny, a module that was previously bundled
    in this distribution, but has since been spun off as a separate
    distribution. Don't worry - it's quick and easy to install.

    At run-time, Type::Tiny also requires the following modules: B,
    B::Deparse, Carp, Data::Dumper, Scalar::Util, Text::Balanced, overload,
    strict and warnings. All of these come bundled with Perl itself. Prior to
    Perl 5.8, Scalar::Util and Text::Balanced do not come bundled with Perl
    and will need installing separately from the CPAN.

    Certain features require additional modules. Tying a variable to a type
    constraint (e.g. `tie my $count, Int`) requires Type::Tie; stack traces on
    exceptions require Devel::StackTrace. The Reply::Plugin::TypeTiny plugin
    for Reply requires Reply (obviously). Devel::LexAlias may *slightly*
    increase the speed of some of Type::Tiny's compiled coderefs.

    Type::Tiny::XS is not required, but if available provides a speed boost
    for some type checks. (Setting the environment variable
    `PERL_TYPE_TINY_XS` to false, or setting `PERL_ONLY` to true will suppress
    the use of Type::Tiny::XS, even if it is available.)

    The test suite additionally requires Test::More, Test::Fatal and
    Test::Requires. Test::More comes bundled with Perl, but if you are using a
    version of Perl older than 5.14, you will need to upgrade to at least
    Test::More version 0.96. Test::Requires and Test::Fatal (plus Try::Tiny
    which Test::Fatal depends on) are bundled with Type::Tiny in the `inc`
    directory, so you do not need to install them separately.

    If using Type::Tiny in conjunction with Moo, then at least Moo 1.000000 is
    recommended. If using Type::Tiny with Moose, then at least Moose 2.0000 is
    recommended. If using Type::Tiny with Mouse, then at least Mouse 1.00 is
    recommended. Type::Tiny is mostly untested against older versions of these
    packages.

TYPE::TINY VERSUS X
  Specio
    Type::Tiny is similar in aim to Specio. The major differences are

    *   Type::Tiny is "tiny" (Specio will eventually have fewer dependencies
        than it currently does, but is unlikely to ever have as few as
        Type::Tiny);

    *   Specio has a somewhat nicer API (better method names; less
        duplication), and its API is likely to improve further. Type::Tiny's
        aims at complete compatibility with current versions of Moose and
        Mouse, so there is a limit to how much I can deviate from the existing
        APIs of (Moose|Mouse)::Meta::TypeConstraint.

  MooseX::Types
    Type::Tiny libraries expose a similar interface to MooseX::Types
    libraries. In most cases you should be able to rewrite a MooseX::Types
    library to use Type::Tiny pretty easily.

  MooX::Types::MooseLike
    Type::Tiny is faster and supports coercions.

  Scalar::Does
    Scalar::Does is somewhat of a precursor to Type::Tiny, but has now been
    rewritten to use Type::Tiny internally.

    It gives you a `does($value, $type)` function that is roughly equivalent
    to `$type->check($value)` except that $type may be one of a list of
    pre-defined strings (instead of a Type::Tiny type constraint); or may be a
    package name in which case it will be assumed to be a role and checked
    with `$value->DOES($type)`.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Type-Tiny>.

SUPPORT
    IRC: support is available through in the *#moops* channel on irc.perl.org
    <http://www.irc.perl.org/channels.html>.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013-2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Type-Tiny
=========

Created:      2013-03-23
Home page:    <https://metacpan.org/release/Type-Tiny>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Type-Tiny>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

1.001_002	2014-10-25

 [ Bug Fixes ]
 - Fix short-circuiting optimizations for parameterized HashRef, ArrayRef,
   ScalarRef, and Map type constraints.
   Fixes RT#99312.
   Marcel Timmerman++
   <https://rt.cpan.org/Ticket/Display.html?id=99312>
 - Inlined version of Types::Standard::Int should check that the value is
   not a reference.

 [ Test Suite ]
 - Fix annoying warning message in test suite with recent versions of
   Exporter::Tiny.

 [ Other ]
 - Make equals/is_a_type_of/is_subtype_of/is_supertype_of in
   Type::Tiny::Union work more like Moose::Meta::TypeConstraint::Union.

1.001_001	2014-09-19

 - Lazy-load Text::Balanced in Type::Parser. (Many parses don't even need
   it.)
 - Lazy-load Type::Tiny::Union in Type::Params.
 - Updated: Prefer Sub::Util over Sub::Name. (The former is smaller.)

1.001_000	2014-09-07

 [ Bug Fixes ]
 - Fix for Type::Registry::DWIM.
   Fixes RT#98458.
   Marcel Montes++
   <https://rt.cpan.org/Ticket/Display.html?id=98458>
 - Fix issues with coercions and native attribute traits with some oldish
   versions of Moose on oldish versions of Perl.
   Fixes RT#98159.
   Peter Flanigan++
   <https://rt.cpan.org/Ticket/Display.html?id=98159>

 [ Documentation ]
 - Updated NEWS file.
 - Updated TODO file.
 - Updates to Type::Tiny::Manual::UsingWithMoose,
   Type::Tiny::Manual::UsingWithMoo, and
   Type::Tiny::Manual::UsingWithMouse.

 [ Test Suite ]
 - Make some of the test case skip_all bits more ambitious; test older
   versions of Moose and Moo than we were testing before.

 [ Other ]
 - Added: Type::Params now provides `compile_named` and `validate_named`
   functions which do the same thing as `compile` and `validate` but are
   better for named arguments.
 - Updated: If Sub::Name is unavailable, but the shiny new core Sub::Util
   is available, then use it instead.
 - Updated: Want Type::Tiny::XS 0.011.
 - `Type::Utils::dwim_type` now allows more control over fallback
   behaviours.

1.000005	2014-10-25

 [ Bug Fixes ]
 - Fix short-circuiting optimizations for parameterized HashRef, ArrayRef,
   ScalarRef, and Map type constraints.
   Fixes RT#99312.
   Marcel Timmerman++
   <https://rt.cpan.org/Ticket/Display.html?id=99312>

 [ Test Suite ]
 - Fix annoying warning message in test suite with recent versions of
   Exporter::Tiny.

1.000004	2014-09-02

 [ Bug Fixes ]
 - Fix for Type::Registry::DWIM.
   Fixes RT#98458.
   Marcel Montes++
   <https://rt.cpan.org/Ticket/Display.html?id=98458>
 - Fix issues with coercions and native attribute traits with some oldish
   versions of Moose on oldish versions of Perl.
   Fixes RT#98159.
   Peter Flanigan++
   <https://rt.cpan.org/Ticket/Display.html?id=98159>

1.000003	2014-08-28

 [ Bug Fixes ]
 - Fix for coercions to parameterized Dict including a mixture of
   inlineable and non-inlineable optional values.
   Fixes RT#98362.
   Gianni Ceccarelli++
   <https://rt.cpan.org/Ticket/Display.html?id=98362>

1.000002	2014-08-18

 [ Bug Fixes ]
 - Fix for overloaded operation fallback on Perl 5.10.x.
   Fixes RT#98113.
   Dagfinn Ilmari Mannsåker++
   Peter Flanigan++
   <https://rt.cpan.org/Ticket/Display.html?id=98113>

1.000001	2014-08-18

 [ Bug Fixes ]
 - Changes to dwim_type() in 0.047_09 broke the fallback to creating class
   types in some circumstances. This broke the MooX::late test suite, and
   some usages of MooX::late.
   Shlomi Fish++

1.000000	2014-08-16	Happy CPAN Day!

 [ Documentation ]
 - Document the Type::Tiny versioning policy.
 - Updated NEWS file.
 - Updated TODO file.

0.047_09	2014-08-12

 [ Bug Fixes ]
 - Fix documentation typo.
   Benct Philip Jonsson++

 [ Documentation ]
 - Improvements and clarifications to Type::Coercion documentation.

 [ Test Suite ]
 - Add tests for Error::TypeTiny::Assertion's predicate methods.
 - Add tests for Type::Coercion's i_really_want_to_unfreeze method.

 [ Other ]
 - Better integration between Type::Library and Type::Registry. If you
   import a type constraint from a Type::Library-based module, it will
   automatically be inserted into your modules' type registry.
 - The to_Foo functions exported by Type::Library-based modules are now
   significantly faster. (But only if the type's coercion is frozen.)
 - Type::Utils::dwim_type now takes into account what OO framework the
   caller is using (Moose or Mouse) if it needs to fall back to asking the
   OO framework about a type constraint.

0.047_08	2014-08-05	Sanity++

 [ Bug Fixes ]
 - Fix non-inlined version of NegativeOrZeroInt.

 [ Documentation ]
 - Document the remaining RT#93345-related issues - see "Deep Caveat" in
   Type::Tiny::Manual::Coercions.
   <http://purl.org/NET/cpan-uri/rt/ticket/93345>

 [ Test Suite ]
 - Alter the Types::Common::* test cases to use Test::TypeTiny.
 - Test to avoid the "too few arguments for type constraint check
   functions" error.
   Diab Jerius++
   <http://purl.org/NET/cpan-uri/rt/ticket/97684>
 - Update t/30-integration/Moose/coercion.t to Moose 2.1200 which throws an
   exception rather than printing a warning when coerce=>1 is used on a
   type constraint with no coercions.

 [ Other ]
 - Prevent inlining of coercions if they've not been frozen.
   <http://purl.org/NET/cpan-uri/rt/ticket/93345>

0.047_07	2014-08-04

 [ Bug Fixes ]
 - The extract_type function was missing from Type::Parser's @EXPORT_OK
   list.

 [ Test Suite ]
 - Test cases for Type::Parser::extract_type.

 [ Other ]
 - Parameterized Maybe constraints are now XS-accelerated.

0.047_06	2014-07-31	What made the Queen go all ice crazy?

 [ Bug Fixes ]
 - Bugfix in coercion inheritance where the child's type_coercion_map
   arrayref would end up as a reference to the parent's type_coercion_map.
   (Which was not good.)

 [ Test Suite ]
 - More Type::Registry test cases.


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "tiny, yet Moo(se)-compatible type constraint",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 1,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.023, CPAN::Meta::Converter version 2.142690",
  "keywords"       => [
                        "Argument Checking",
                        "Argument Validation",
                        "Moo",
                        "Moose",
                        "Mouse",
                        "Parameter Checking",
                        "Parameter Validation",
                        "Schema",
                        "Type Coercion",
                        "Type Constraint",
                        "Type Library",
                        "Validation",
                      ],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Type-Tiny",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => {
                                       recommends => { "CPAN::Meta::Requirements" => "2.000" },
                                       requires   => { "ExtUtils::MakeMaker" => 6.17 },
                                     },
                        develop   => { suggests => { "Dist::Inkt::Profile::TOBYINK" => 0 } },
                        runtime   => {
                                       conflicts  => { "Kavorka" => "<= 0.013", "Types::ReadOnly" => "<= 0.001" },
                                       recommends => {
                                                       "Devel::LexAlias" => 0.05,
                                                       "Devel::StackTrace" => 0,
                                                       "perl" => 5.010001,
                                                       "Sub::Name" => 0,
                                                       "Type::Tie" => 0,
                                                       "Type::Tiny::XS" => 0.011,
                                                     },
                                       requires   => { "Exporter::Tiny" => 0.026, "perl" => 5.006001 },
                                       suggests   => { Moo => "1.000000", Moose => "2.0000", Mouse => "1.00", Reply => 0 },
                                     },
                        test      => {
                                       recommends => { "Test::Tester" => 0.109, "Test::Warnings" => 0 },
                                       requires   => { "Test::More" => 0.96 },
                                     },
                      },
  "provides"       => {
                        "Devel::TypeTiny::Perl56Compat"            => {
                                                                        file => "lib/Devel/TypeTiny/Perl56Compat.pm",
                                                                        version => "1.001_002",
                                                                      },
                        "Devel::TypeTiny::Perl58Compat"            => {
                                                                        file => "lib/Devel/TypeTiny/Perl58Compat.pm",
                                                                        version => "1.001_002",
                                                                      },
                        "Error::TypeTiny"                          => { file => "lib/Error/TypeTiny.pm", version => "1.001_002" },
                        "Error::TypeTiny::Assertion"               => { file => "lib/Error/TypeTiny/Assertion.pm", version => "1.001_002" },
                        "Error::TypeTiny::Compilation"             => { file => "lib/Error/TypeTiny/Compilation.pm", version => "1.001_002" },
                        "Error::TypeTiny::WrongNumberOfParameters" => {
                                                                        file => "lib/Error/TypeTiny/WrongNumberOfParameters.pm",
                                                                        version => "1.001_002",
                                                                      },
                        "Eval::TypeTiny"                           => { file => "lib/Eval/TypeTiny.pm", version => "1.001_002" },
                        "Reply::Plugin::TypeTiny"                  => { file => "lib/Reply/Plugin/TypeTiny.pm", version => "1.001_002" },
                        "Test::TypeTiny"                           => { file => "lib/Test/TypeTiny.pm", version => "1.001_002" },
                        "Type::Coercion"                           => { file => "lib/Type/Coercion.pm", version => "1.001_002" },
                        "Type::Coercion::FromMoose"                => { file => "lib/Type/Coercion/FromMoose.pm", version => "1.001_002" },
                        "Type::Coercion::Union"                    => { file => "lib/Type/Coercion/Union.pm", version => "1.001_002" },
                        "Type::Library"                            => { file => "lib/Type/Library.pm", version => "1.001_002" },
                        "Type::Params"                             => { file => "lib/Type/Params.pm", version => "1.001_002" },
                        "Type::Parser"                             => { file => "lib/Type/Parser.pm", version => "1.001_002" },
                        "Type::Registry"                           => { file => "lib/Type/Registry.pm", version => "1.001_002" },
                        "Type::Tiny"                               => { file => "lib/Type/Tiny.pm", version => "1.001_002" },
                        "Type::Tiny::Class"                        => { file => "lib/Type/Tiny/Class.pm", version => "1.001_002" },
                        "Type::Tiny::Duck"                         => { file => "lib/Type/Tiny/Duck.pm", version => "1.001_002" },
                        "Type::Tiny::Enum"                         => { file => "lib/Type/Tiny/Enum.pm", version => "1.001_002" },
                        "Type::Tiny::Intersection"                 => { file => "lib/Type/Tiny/Intersection.pm", version => "1.001_002" },
                        "Type::Tiny::Role"                         => { file => "lib/Type/Tiny/Role.pm", version => "1.001_002" },
                        "Type::Tiny::Union"                        => { file => "lib/Type/Tiny/Union.pm", version => "1.001_002" },
                        "Type::Utils"                              => { file => "lib/Type/Utils.pm", version => "1.001_002" },
                        "Types::Common::Numeric"                   => { file => "lib/Types/Common/Numeric.pm", version => "1.001_002" },
                        "Types::Common::String"                    => { file => "lib/Types/Common/String.pm", version => "1.001_002" },
                        "Types::Standard"                          => { file => "lib/Types/Standard.pm", version => "1.001_002" },
                        "Types::Standard::ArrayRef"                => { file => "lib/Types/Standard/ArrayRef.pm", version => "1.001_002" },
                        "Types::Standard::Dict"                    => { file => "lib/Types/Standard/Dict.pm", version => "1.001_002" },
                        "Types::Standard::HashRef"                 => { file => "lib/Types/Standard/HashRef.pm", version => "1.001_002" },
                        "Types::Standard::Map"                     => { file => "lib/Types/Standard/Map.pm", version => "1.001_002" },
                        "Types::Standard::ScalarRef"               => { file => "lib/Types/Standard/ScalarRef.pm", version => "1.001_002" },
                        "Types::Standard::Tuple"                   => { file => "lib/Types/Standard/Tuple.pm", version => "1.001_002" },
                        "Types::TypeTiny"                          => { file => "lib/Types/TypeTiny.pm", version => "1.001_002" },
                      },
  "release_status" => "testing",
  "resources"      => {
                        bugtracker   => { web => "http://rt.cpan.org/Dist/Display.html?Queue=Type-Tiny" },
                        homepage     => "https://metacpan.org/release/Type-Tiny",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-type-tiny.git",
                                          web  => "https://github.com/tobyink/p5-type-tiny",
                                        },
                        x_identifier => "http://purl.org/NET/cpan-uri/dist/Type-Tiny/project",
                        x_IRC        => "irc://irc.perl.org/#moops",
                      },
  "version"        => "1.001_002",
  "x_breaks"       => { "Kavorka" => "<= 0.013", "Types::ReadOnly" => "<= 0.001" },
  "x_contributors" => [
                        "Diab Jerius (DJERIUS) <djerius\@cpan.org>",
                        "Vyacheslav Matyukhin (MMCLERIC) <mmcleric\@cpan.org>",
                        "Peter Flanigan (PJFL) <pjfl\@cpan.org>",
                        "Pierre Masci",
                        "Mark Stosberg (MARKSTOS) <markstos\@cpan.org>",
                        "David Steinbrunner <dsteinbrunner\@pobox.com>",
                        "Samuel Kaufman (SKAUFMAN) <skaufman\@cpan.org>",
                        "Graham Knop (HAARG) <haarg\@cpan.org>",
                        "Peter Karman (KARMAN) <karman\@cpan.org>",
                        "Alexander Hartmaier (ABRAXXA) <abraxxa\@cpan.org>",
                        "Dagfinn Ilmari Manns\xE5ker (ILMARI) <ilmari\@cpan.org>",
                        "Gianni Ceccarelli (DAKKAR) <dakkar\@cpan.org>",
                      ],
};

my %dynamic_config;
do {
$meta->{prereqs}{runtime}{requires}{'Scalar::Util'}   = '1.13'  if $] < 5.007003;
$meta->{prereqs}{runtime}{requires}{'Data::Dumper'}   = '1.121' if $] < 5.008001;
$meta->{prereqs}{runtime}{requires}{'Text::Balanced'} = '1.95'  if $] < 5.007003;
$meta->{prereqs}{runtime}{requires}{'Exporter'}       = '5.57'  if $] < 5.009001;

if ($ENV{EXTENDED_TESTING} and $meta->{version} =~ /_/)
{
	$meta->{prereqs}{test}{requires}{'Moose'} = '2.0600';
}

};
for my $stage (keys %{$meta->{prereqs}})
{
	my $conflicts = $meta->{prereqs}{$stage}{conflicts} or next;
	eval { require CPAN::Meta::Requirements } or last;
	$conflicts = 'CPAN::Meta::Requirements'->from_string_hash($conflicts);
	
	for my $module ($conflicts->required_modules)
	{
		eval "require $module" or next;
		my $installed = eval(sprintf('$%s::VERSION', $module));
		$conflicts->accepts_module($module, $installed) or next;
		
		my $message = "\n".
			"** This version of $meta->{name} conflicts with the version of\n".
			"** module $module ($installed) you have installed.\n";
		die($message . "\n" . "Bailing out")
			if $stage eq 'build' || $stage eq 'configure';
		
		$message .= "**\n".
			"** It's strongly recommended that you update it after\n".
			"** installing this version of $meta->{name}.\n";
		warn("$message\n");
	}
}

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t t/20-unit/Devel-TypeTiny-Perl56Compat/*.t t/20-unit/Devel-TypeTiny-Perl58Compat/*.t t/20-unit/Error-TypeTiny-Assertion/*.t t/20-unit/Error-TypeTiny-Compilation/*.t t/20-unit/Error-TypeTiny-WrongNumberOfParameters/*.t t/20-unit/Error-TypeTiny/*.t t/20-unit/Eval-TypeTiny/*.t t/20-unit/Test-TypeTiny/*.t t/20-unit/Type-Coercion-FromMoose/*.t t/20-unit/Type-Coercion-Union/*.t t/20-unit/Type-Coercion/*.t t/20-unit/Type-Library/*.t t/20-unit/Type-Params/*.t t/20-unit/Type-Parser/*.t t/20-unit/Type-Registry/*.t t/20-unit/Type-Tiny-Class/*.t t/20-unit/Type-Tiny-Duck/*.t t/20-unit/Type-Tiny-Enum/*.t t/20-unit/Type-Tiny-Intersection/*.t t/20-unit/Type-Tiny-Role/*.t t/20-unit/Type-Tiny-Union/*.t t/20-unit/Type-Tiny-_HalfOp/*.t t/20-unit/Type-Tiny/*.t t/20-unit/Type-Utils/*.t t/20-unit/Types-Common-Numeric/*.t t/20-unit/Types-Common-String/*.t t/20-unit/Types-Standard/*.t t/20-unit/Types-TypeTiny/*.t t/30-integration/Class-InsideOut/*.t t/30-integration/Exporter-Tiny/*.t t/30-integration/Function-Parameters/*.t t/30-integration/Kavorka/*.t t/30-integration/Moo/*.t t/30-integration/Moops/*.t t/30-integration/Moose/*.t t/30-integration/MooseX-Getopt/*.t t/30-integration/MooseX-Types/*.t t/30-integration/Mouse/*.t t/30-integration/MouseX-Types/*.t t/30-integration/Object-Accessor/*.t t/30-integration/Return-Type/*.t t/30-integration/Sub-Quote/*.t t/30-integration/Switcheroo/*.t t/30-integration/Type-Tie/*.t t/30-integration/Validation-Class-Simple/*.t t/30-integration/match-simple/*.t t/40-regression/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

